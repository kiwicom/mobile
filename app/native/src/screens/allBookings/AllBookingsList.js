// @flow

import * as React from 'react';
import { View, Text } from 'react-native';
import { createFragmentContainer, graphql } from 'react-relay';
import idx from 'idx';

import BookingOverviewRow from '../../components/bookings/OverviewRow';
import AllBookingsListRowError from './AllBookingsListRowError';
import AllBookingsAssetsDownloader from './AllBookingsAssetsDownloader';
import Large from '../../components/visual/text/Large';

import type { AllBookingsList_bookings } from './__generated__/AllBookingsList_bookings.graphql';
import type { Navigation } from '../../types/Navigation';

type Props = {|
  bookings: AllBookingsList_bookings,
  navigation: Navigation,
|};

type State = {|
  loading: boolean,
|};

const filterEdges = allEdges => {
  const filteredTrips = [[], []];
  allEdges.forEach(edge => {
    const departureLocalTime = idx(edge, _ => _.node.departure.localTime);
    if (departureLocalTime) {
      if (new Date(departureLocalTime).getTime() >= Date.now()) {
        // future trips
        filteredTrips[0].push(edge);
      }
      filteredTrips[1].push(edge);
    }
  });
  return filteredTrips;
};

/**
 * Renders past and future trips. It automatically downloads all assets like
 * plane tickets and invoices for future trips (not for the past trips).
 */
export class AllBookingsListWithoutData extends React.Component<Props, State> {
  render = () => {
    const allEdges = idx(this.props, _ => _.bookings.allBookings.edges) || [];
    const [futureEdges, pastEdges] = filterEdges(allEdges);

    const title = titleString => (
      <View style={{ paddingTop: 25, paddingBottom: 10 }}>
        <Large>{titleString}</Large>
      </View>
    );

    const renderEdges = (edges, downloadAssets: boolean = false) => {
      const listLength = edges.length;
      return listLength > 0 ? (
        edges.map((edge, index) => {
          if (edge) {
            const { node, cursor } = edge;
            return [
              downloadAssets && (
                <AllBookingsAssetsDownloader
                  data={node && node.assets}
                  key="downloader"
                />
              ),
              <BookingOverviewRow
                node={node}
                key={cursor}
                onPress={() =>
                  this.props.navigation.navigate('SingleBooking', {
                    bookingId: node && node.id,
                    bookingDatabaseId: node && node.databaseId,
                  })
                }
                showSeparator={listLength !== index + 1} // everywhere except the last row in the list
              />,
            ];
          } else {
            return <AllBookingsListRowError />;
          }
        })
      ) : (
        <View style={{ paddingTop: 20, paddingBottom: 20 }}>
          <Text>No trips yet.</Text>
        </View>
      );
    };

    return (
      <View>
        {title('Future trips')}
        {renderEdges(futureEdges, true)}

        {title('Past trips')}
        {renderEdges(pastEdges)}
      </View>
    );
  };
}

export default createFragmentContainer(
  AllBookingsListWithoutData,
  graphql`
    fragment AllBookingsList_bookings on RootQuery {
      allBookings {
        edges {
          cursor
          node {
            id
            databaseId
            assets {
              ...AllBookingsAssetsDownloader
            }
            departure {
              localTime
            }
            ...OverviewRow_node
          }
        }
      }
    }
  `,
);
