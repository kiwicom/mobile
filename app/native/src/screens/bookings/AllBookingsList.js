// @flow

import * as React from 'react';
import { View, Text } from 'react-native';
import { createFragmentContainer, graphql } from 'react-relay';
import idx from 'idx';

import BookingListRow from './AllBookingsListRow';
import BookingListRowError from './AllBookingsListRowError';
import AssetsDownloader from '../../services/assets/AssetsDownloader';

import type { AllBookingsList_bookings } from './__generated__/AllBookingsList_bookings.graphql';
import type { Navigation } from '../../types/Navigation';

type Props = {
  bookings: AllBookingsList_bookings,
  navigation: Navigation,
};

type State = {
  loading: boolean,
};

export class BookingsListWithoutData extends React.Component<Props, State> {
  render = () => {
    const edges = idx(this.props, _ => _.bookings.allBookings.edges);
    return (
      <View>
        {edges ? (
          edges.map(edge => {
            if (edge) {
              const { node, cursor } = edge;

              if (node && node.assets) {
                const { ticketUrl, invoiceUrl } = node.assets;
                if (ticketUrl) {
                  AssetsDownloader({ url: ticketUrl });
                }
                if (invoiceUrl) {
                  AssetsDownloader({ url: invoiceUrl });
                }
              }

              return (
                <BookingListRow
                  node={node}
                  key={cursor}
                  navigation={this.props.navigation}
                />
              );
            } else {
              return <BookingListRowError />;
            }
          })
        ) : (
          <Text>We couldn&apos;t load the bookings (missing data).</Text>
        )}
      </View>
    );
  };
}

export default createFragmentContainer(
  BookingsListWithoutData,
  graphql`
    fragment AllBookingsList_bookings on RootQuery {
      allBookings {
        edges {
          cursor
          node {
            assets {
              ticketUrl
              invoiceUrl
            }
            ...AllBookingsListRow_node
          }
        }
      }
    }
  `,
);
