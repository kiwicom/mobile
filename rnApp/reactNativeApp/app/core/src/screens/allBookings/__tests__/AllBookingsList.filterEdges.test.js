// @flow

import { filterEdges } from '../AllBookingsList';

// This is our pivot: dates after and equal to this date are considered "future"
// $FlowExpectedError: Flow is not happy about this magic, but it's OK
Date.now = jest.fn().mockReturnValue(new Date(3));

const dataset = [
  { node: { arrival: { localTime: new Date(1) }, assets: null }, cursor: 'x' },
  { node: { arrival: { localTime: new Date(2) }, assets: null }, cursor: 'x' },
  { node: { arrival: { localTime: new Date(3) }, assets: null }, cursor: 'x' }, // future
  { node: { arrival: { localTime: new Date(4) }, assets: null }, cursor: 'x' }, // future
  { node: { arrival: { localTime: new Date(5) }, assets: null }, cursor: 'x' }, // future
];

it('filters future/current and past trips based on arrival', () => {
  expect(filterEdges(dataset)).toMatchSnapshot();
});
