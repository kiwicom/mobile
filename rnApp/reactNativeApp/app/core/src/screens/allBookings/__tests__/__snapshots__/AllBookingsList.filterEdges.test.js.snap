// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters future/current and past trips based on arrival 1`] = `
Array [
  Array [
    Object {
      "cursor": "x",
      "node": Object {
        "arrival": Object {
          "localTime": 1970-01-01T00:00:00.003Z,
        },
        "assets": null,
      },
    },
    Object {
      "cursor": "x",
      "node": Object {
        "arrival": Object {
          "localTime": 1970-01-01T00:00:00.004Z,
        },
        "assets": null,
      },
    },
    Object {
      "cursor": "x",
      "node": Object {
        "arrival": Object {
          "localTime": 1970-01-01T00:00:00.005Z,
        },
        "assets": null,
      },
    },
  ],
  Array [
    Object {
      "cursor": "x",
      "node": Object {
        "arrival": Object {
          "localTime": 1970-01-01T00:00:00.001Z,
        },
        "assets": null,
      },
    },
    Object {
      "cursor": "x",
      "node": Object {
        "arrival": Object {
          "localTime": 1970-01-01T00:00:00.002Z,
        },
        "assets": null,
      },
    },
  ],
]
`;
